src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp

```diff
@@ -5311,6 +5313,365 @@ class StubGenerator: public StubCodeGenerator {
+    const FloatRegister vmul0 = v4, vmul1 = v5, vmul2 = v6, vmul3 = v7;
+    const FloatRegister vpow = v8;  // powers of 31: <31^3, ..., 31^0>
+    const FloatRegister vpowm = v9;
+
+    assert_different_registers(ary, cnt, result);
+    assert_different_registers(vdata0, vdata1, vdata2, vdata3, vhalf0, vhalf1, vhalf2, vhalf3,
+                               vmul0, vmul1, vmul2, vmul3, vpow, vpowm);
+
+    Label SMALL_LOOP, LARGE_LOOP_PREHEADER, LARGE_LOOP, TAIL, TAIL_SHORTCUT, RELATIVE;
+
+    // Vectorization factor
+    const size_t vf = eltype == T_BOOLEAN || eltype == T_BYTE ? 8
+                      : eltype == T_CHAR || eltype == T_SHORT ? 8
+                      : eltype == T_INT                       ? 4
+                                                              : 0;
+    guarantee(vf, "unsupported eltype");
+
+    // Unroll factor
+    const size_t uf = 4;
+
+    // Effective vectorization factor
+    const size_t evf = vf * uf;
+
+    __ align(CodeEntryAlignment);
+
+    const char *mark_name = "";
+    switch (eltype) {
+    case T_BOOLEAN:
+      mark_name = "_large_arrays_hashcode_boolean";
+      break;
+    case T_BYTE:
+      mark_name = "_large_arrays_hashcode_byte";
+      break;
+    case T_CHAR:
+      mark_name = "_large_arrays_hashcode_char";
+      break;
+    case T_SHORT:
+      mark_name = "_large_arrays_hashcode_short";
+      break;
+    case T_INT:
+      mark_name = "_large_arrays_hashcode_int";
+      break;
+    default:
+      mark_name = "_large_arrays_hashcode_incorrect_type";
+      __ should_not_reach_here();
+    };
+
+    StubCodeMark mark(this, "StubRoutines", mark_name);
+
+    address entry = __ pc();
+    __ enter();
+
+    Assembler::SIMD_Arrangement load_arrangement =
+        eltype == T_BOOLEAN || eltype == T_BYTE ? Assembler::T8B
+        : eltype == T_CHAR || eltype == T_SHORT ? Assembler::T8H
+        : eltype == T_INT                       ? Assembler::T4S
+                                                : Assembler::INVALID_ARRANGEMENT;
+    guarantee(load_arrangement != Assembler::INVALID_ARRANGEMENT, "invalid arrangement");
+
+    const int multiply_by_halves =
+        load_arrangement == Assembler::T4S || load_arrangement == Assembler::T4H   ? false
+        : load_arrangement == Assembler::T8B || load_arrangement == Assembler::T8H ? true
+                                                                                   : -1;
+    guarantee(multiply_by_halves != -1, "unknown multiplication algorithm");
+
+    const int small_loop_size = load_arrangement == Assembler::T4S   ? 20  // 5 insts
+                                : load_arrangement == Assembler::T4H ? 24  // 6 insts
+                                : load_arrangement == Assembler::T8H ? 36  // 9 insts
+                                : load_arrangement == Assembler::T8B ? 40  // 10 insts
+                                                                     : -1; // invalid
```

Comment:
This is extremely fragile in the presence of code change. Can we not simply delete it?

Reply:
There's a `guarantee()` at the end of the loop to verify the size so the code change shouldn't be left unnoticed.

Reply:
Removed by https://github.com/openjdk/jdk/pull/18487/commits/9f4ae8554106d3e1290ee127b848d7f48952e5fb .